### PR TITLE
[core][autoscaler] Autoscaler doesn't scale up correctly when the KubeRay RayCluster is not in the goal state

### LINKED ISSUE #46473: [Ray autoscaler v2] Can't scaler up when using autoscaler v2
### What happened + What you expected to happen

For the same environment, only changing the use of autoscaler v1 or v2, for a one-time submission of 8000 tasks, v1 can work normally, but v2 will always be stuck, can not be scaled up.
Also, I want to know what has made recent progress in AutoScaler V2? It seems that it has not been updated for a long time.

### Versions / Dependencies

Ray 2.23.0
Kuberay 1.1.1

### Reproduction script

```
import ray
import time
import os
import random

@ray.remote(max_retries=5, num_cpus=8)
def inside_ray_task():
    sleep_time = random.randint(120, 600)

    start_time = time.perf_counter()
    while True:
        if(time.perf_counter() - start_time < sleep_time):
            time.sleep(0.001)
        else:
            break
  
@ray.remote(max_retries=0)
def outside_ray_task():
    future_list = []
    for i in range(8000):
        future_list.append(inside_ray_task.remote())
    ray.get(future_list)

if __name__ == '__main__':
    ray.init("ray://localhost:10001")
    ray.get(outside_ray_task.remote())
```
3adc4197-8928-4f55-9bce-a332d21b3b07

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by kevin85421**
### Issue
  * Create a Autoscaler V2 RayCluster CR.
    * head Pod: `num-cpus: 0`
    * worker Pod: Each worker Pod has 1 CPU, and the `maxReplicas` of the worker group is 10.
  * Run the following script in the head Pod: https://gist.github.com/kevin85421/6f09368ba48572e28f53654dca854b57
  * There are 10 scale requests to add a new node. However, only some of them will be created (e.g., 5).

### Reason

In the reproduction script above, the `cloud_instance_updater` will send a request to scale up one worker Pod 10 times because the `maxReplicas` of the worker group is set to 10.

However, the construction of the scale_request depends on the Pods in the Kubernetes cluster. For example, 

* cluster state: RayCluster Replicas: 2, Ray Pods: 1
  * 1st scale request: launch 1 node --> goal state: RayCluster Replicas: 2 (Ray Pods + 1)
  * 2nd scale request: launch 1 node --> goal state: RayCluster Replicas: 2 (Ray Pods + 1) --> **this should be 3!**

The above example is expected to create 3 Pods. However, it will ultimately create only 2 Pods.

### Solution

Use RayCluster CR instead of Ray Pods to build scale requests.

## Related issue number

Closes #46473

## Checks

10 worker Pods are created successfully.

Screenshot 2024-11-24 at 2 11 39 AM

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py [hunk 1]
```diff
@@ -422,14 +432,19 @@ def _get_workers_groups_with_deletes(
...
+                worker_to_delete_set.add(worker)
                 if worker in node_set:
                     worker_groups_with_pending_deletes.add(node_type)
                     break
 
         worker_groups_with_finished_deletes = (
             worker_groups_with_deletes - worker_groups_with_pending_deletes
         )
-        return worker_groups_with_pending_deletes, worker_groups_with_finished_deletes
+        return (
```
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:443]: nit: function signature as well.

File: python/ray/autoscaler/v2/tests/test_node_provider.py [hunk 2]
```diff
@@ -492,6 +492,43 @@ def test_pending_deletes(self):
...
+
+        patches = self.mock_client.get_patches(
+            f"rayclusters/{self.provider._cluster_name}"
+        )
+        assert len(patches) == 1
+        assert patches[0] == {
+            "op": "replace",
+            "path": "/spec/workerGroupSpecs/0/replicas",
+            "value": desired_replicas + 1,
+        }
```
Comment (rickyyx) [inline python/ray/autoscaler/v2/tests/test_node_provider.py:530]: nit: is it possible to also add tests that to delete workers are handled correctly when calculating the goal state by the change for regression?

**Review by rickyyx — APPROVED** [review]
Ah, this makes muc sense. We could merge this than https://github.com/ray-project/ray/pull/47967 then?

**Comment by kevin85421** [comment]
> We could merge this than #47967 then?

Oops, I didn't realize that there is a similar PR for #46473. I think merging this PR is sufficient. This PR is more ready for merging from both testing and KubeRay perspectives (as it relies on some outdated assumptions of KubeRay). In addition, Autoscaler V2 is currently on the critical path for the KubeRay release. I prefer to go with this PR and I will communicate with the contributor (perhaps offering them other issues if they are interested in contributing).

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py (hunk 1 above)
Comment (kevin85421) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:443]: rename the function?

File: python/ray/autoscaler/v2/tests/test_node_provider.py (hunk 2 above)
Comment (kevin85421) [reply python/ray/autoscaler/v2/tests/test_node_provider.py:530]: added https://github.com/ray-project/ray/pull/48909/commits/24ba62a6b198f1e2d2ee9ceb029e2bbb612d3849

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py (hunk 1 above)
Comment (kevin85421) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:441]: renamed the function: https://github.com/ray-project/ray/pull/48909/commits/24ba62a6b198f1e2d2ee9ceb029e2bbb612d3849

Comment (kevin85421) [reply python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py:441]: good catch! updated https://github.com/ray-project/ray/pull/48909/commits/2c9c3d0afe84f5f1c727b435960206399d573eb7

**Pull request merged**